### ISSUE TITLE
Allow implicit returns in CEs

### ISSUE THREAD

**Issue by Dimension4** [fsharp/fslang-suggestions#1115]
# Allow implicit returns in CEs

I propose we allow `return` to be implicit on the last expression in a CE that supports returns. Existing behavior of implicit `yield`s wouldn't change: if a CE supports `yield` use that, else use `return`. If neither is supported, keep existing behavior. The behavior of `return!` is also untouched.

Like `seq { 1 }` is of type `Seq<int>`, `async { 1 }` would then be of type `Async<int>` instead of `Async<unit>`.

### Pros
 - Consistency with the behavior outside of CEs.

### Cons
- Could cause further confusion between `return` and `return!`, especially in recursive functions, like:
  ```fs
    let rec foo () = 
        foo ()

    let rec asyncFoo () = 
        async {
            asyncFoo ()
        }
  ```
  Although that case already causes confusion with the explicit `return`.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): I'm not familiar with the compiler, but I guess this shouldn't be too complicated: S.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by cartermp** [comment]
Which would be used if:

* `Yield` and `Return` are defined
* `Yield and `ReturnFrom` are defined
* `Return and `ReturnFrom` are defined

**Comment by vzarytovskii** [comment]
> Which would be used if:
> 
> * `Yield` and `Return` are defined
> * `Yield and `ReturnFrom` are defined
> * `Return and `ReturnFrom` are defined

Yeah, if we will define a priority (for example, we use Yield if defined, then Return), there theoretically may be confusing situations like:
If we have only `Return` defined, for the following builder
```fsharp
builder {
    2
}
```
 `Returm` will be used, then if we define `Yield`, it will change behaviour all of a sudden.

**Comment by cartermp** [comment]
@Dimension4 which discussions and where?

Also: why would `Yield` be preferred over `Return`?

**Comment by drhumlen** [comment]
How common is it for a CE to have both yield and return though? I don't think I've ever seen it.

But in that rare case, I think you should be forced to be explicit for both yield and return. But defaulting to yield to avoid breaking existing behavior is interesting too.

'return!' should be explicit just like 'yield!'

**Comment by edgarfgp** [comment]
@dsyme Any thoughts on this suggestion? . This will definitely help us in Fabulous to take our DSL to the next level for V3 😀 cc @ TimLariviere

**Comment by TimLariviere** [comment]
A use-case for it came up while trying to improve the DSL of Fabulous.

We currently represent parent-children relationship in UI in 2 different ways
```fs
// Single child -- passed as a method argument
Page(
    Label("Hello")
)

// Multiple children -- passed through implicit yield of a computation expression
VStack() {
    Label("Hello")
    Label("World")
}
```

While I think this is fine, I believe it would be better if we always represent parent-child relationship the same way through computation expressions. e.g.
```fs
Page() {
    VStack() {
        Label("Hello")
        Label("World")
    }
}
```

But today only implicit yields are supported. We could implement a CE selecting only the last yielded child, but the compiler is no longer giving a warning to tell the user that previous children will be discarded.

The ability to use implicit return would allow Fabulous to achieve this.
Single child CE would only implement `Return` and Multiple children CE would only implement `Yield`.

**Comment by dsyme** [comment]
Implicit returns do make a lot of sense, to match explicit yields

Like implicit ignores, the main problem is unit types, e.g

```
async { 1 }
```

This is valid code today:
```fsharp
> async { 1 };;

  async { 1 };;
  --------^

stdin(1,9): warning FS0020: The result of this expression has type 'int' and is implicitly ignored. Consider using 'ignore' to discard this value explicitly, e.g. 'expr |> ignore', or 'let' to bind the result to a name, e.g. 'let result = expr'.

val it: Async<unit> =
  Microsoft.FSharp.Control.FSharpAsync`1[Microsoft.FSharp.Core.Unit]
```
